core/src/software/aws/toolkits/core/credentials/sso/SsoAccessTokenProvider.kt

```diff
@@ -0,0 +1,135 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package software.aws.toolkits.core.credentials.sso
+
+import kotlinx.coroutines.delay
+import software.amazon.awssdk.services.ssooidc.SsoOidcClient
+import software.amazon.awssdk.services.ssooidc.model.AuthorizationPendingException
+import software.amazon.awssdk.services.ssooidc.model.InvalidClientException
+import software.amazon.awssdk.services.ssooidc.model.SlowDownException
+import java.time.Clock
+import java.time.Duration
+import java.time.Instant
+
+/**
+ * Takes care of creating/refreshing the SSO access token required to fetch SSO-based credentials.
+ */
+class SsoAccessTokenProvider(
+    private val ssoUrl: String,
+    private val ssoRegion: String,
+    private val onPendingToken: SsoLoginCallback,
+    private val cache: SsoCache,
+    private val client: SsoOidcClient,
+    private val clock: Clock = Clock.systemUTC()
```

Comment:
should we pull the clock into a constant for all the SSO-related stuff? can't think of a reason why they would ever be different

Reply:
I'd rather have it hidden by default so callers shouldn't know about it.

Reply:
I am not sure its worth it. Frankly we really only have it due to testing >.>